File: backend/app/models/token_configuration.py
from decimal import Decimal
from typing import List, Optional
from sqlalchemy import String, Numeric, Integer, JSON
from sqlalchemy.orm import Mapped, mapped_column, relationship

from .base import Base


class TokenConfiguration(Base):
    """Model for storing token configuration data"""
    
    __tablename__ = "token_configurations"

    # Basic token information
    creator_wallet: Mapped[str] = mapped_column(String, nullable=False, index=True)
    token_name: Mapped[str] = mapped_column(String, nullable=False)
    token_symbol: Mapped[str] = mapped_column(String, nullable=False)
    icon_url: Mapped[Optional[str]] = mapped_column(String, nullable=True)
    decimals: Mapped[int] = mapped_column(Integer, nullable=False)
    total_supply: Mapped[Decimal] = mapped_column(Numeric, nullable=False)
    
    # Distribution and deployment data
    csv_data: Mapped[Optional[dict]] = mapped_column(JSON, nullable=True)
    fee_paid_tx: Mapped[Optional[str]] = mapped_column(String, nullable=True)
    deployment_status: Mapped[str] = mapped_column(
        String,
        nullable=True,
        server_default="pending"
    )

    # Relationships
    distributions: Mapped[List["TokenDistribution"]] = relationship(
        "TokenDistribution",
        back_populates="token_config",
        cascade="all, delete-orphan"
    )
    deployment_logs: Mapped[List["DeploymentLog"]] = relationship(
        "DeploymentLog",
        back_populates="token_config",
        cascade="all, delete-orphan"
    )
    transfer_transactions: Mapped[List["TransferTransaction"]] = relationship(
        "TransferTransaction",
        back_populates="token_config",
        cascade="all, delete"
    ) 

File: backend/src/models/TokenConfiguration.js
const { DataTypes } = require('sequelize');
const sequelize = require('../db/config');

const TokenConfiguration = sequelize.define('TokenConfiguration', {
  id: {
    type: DataTypes.INTEGER,
    primaryKey: true,
    autoIncrement: true
  },
  creatorWallet: {
    type: DataTypes.STRING,
    allowNull: false,
    field: 'creator_wallet'
  },
  tokenName: {
    type: DataTypes.STRING,
    allowNull: false,
    field: 'token_name'
  },
  tokenSymbol: {
    type: DataTypes.STRING,
    allowNull: false,
    field: 'token_symbol'
  },
  iconUrl: {
    type: DataTypes.STRING,
    field: 'icon_url'
  },
  decimals: {
    type: DataTypes.INTEGER,
    allowNull: false,
    defaultValue: 18
  },
  totalSupply: {
    type: DataTypes.DECIMAL,
    allowNull: false,
    field: 'total_supply'
  },
  distributionsJson: {
    type: DataTypes.JSONB,
    field: 'distributions_json'
  },
  selectedChains: {
    type: DataTypes.JSONB,
    field: 'selected_chains'
  },
  feePaidTx: {
    type: DataTypes.STRING,
    field: 'fee_paid_tx'
  },
  deploymentStatus: {
    type: DataTypes.STRING,
    allowNull: false,
    defaultValue: 'pending',
    field: 'deployment_status'
  },
  deploymentError: {
    type: DataTypes.TEXT,
    field: 'deployment_error'
  }
}, {
  tableName: 'token_configurations',
  timestamps: true,
  createdAt: 'created_at',
  updatedAt: 'updated_at'
});

module.exports = TokenConfiguration; 